XamarinCommunityToolkit/Helpers/BindablePropertyBuilder.shared.cs

```diff
@@ -0,0 +1,180 @@
+		/// <returns>The <see cref="BindablePropertyBuilder"/> to support the fluent syntax</returns>
+		public BindablePropertyBuilder SetPropertyChangedDelegate(BindableProperty.BindingPropertyChangedDelegate propertyChangedDelegate)
+		{
+			bpParameters.BindingPropertyChangedDelegate = propertyChangedDelegate;
+			return this;
+		}
+
+		/// <summary>
+		/// Sets the delegate that runs before a bound value is changed.
+		/// </summary>
+		/// <param name="propertyChangingDelegate">The delegate that runs before a bound value is changed.</param>
+		/// <returns>The <see cref="BindablePropertyBuilder"/> to support the fluent syntax</returns>
+		public BindablePropertyBuilder SetPropertyChangingDelegate(BindableProperty.BindingPropertyChangingDelegate propertyChangingDelegate)
+		{
+			bpParameters.BindingPropertyChangingDelegate = propertyChangingDelegate;
+			return this;
+		}
+
+		/// <summary>
+		/// Sets the delegate used to coerce the range of a value.
+		/// </summary>
+		/// <param name="coerceValueDelegate">A delegate used to coerce the range of a value.</param>
+		/// <returns>The <see cref="BindablePropertyBuilder"/> to support the fluent syntax</returns>
+		public BindablePropertyBuilder SetCoerceValueDelegate(BindableProperty.CoerceValueDelegate coerceValueDelegate)
+		{
+			bpParameters.CoerceValueDelegate = coerceValueDelegate;
+			return this;
+		}
+
+		/// <summary>
+		/// Sets the delegate used to initialize the default value for reference types.
+		/// </summary>
+		/// <param name="createDefaultValueDelegate">The delegate used to initialize the default value for reference types.</param>
+		/// <returns>The <see cref="BindablePropertyBuilder"/> to support the fluent syntax</returns>
+		public BindablePropertyBuilder SetCreateDefaultValueDelegate(BindableProperty.CreateDefaultValueDelegate createDefaultValueDelegate)
+		{
+			bpParameters.CreateDefaultValueDelegate = createDefaultValueDelegate;
+			return this;
+		}
+
+		/// <summary>
+		/// Instantiates the <see cref="BindableProperty"/> based on the values that have been set.
+		/// </summary>
+		/// <returns>An instance of <see cref="BindableProperty"/> with the values that have been set</returns>
+		/// <exception cref="ArgumentNullException">
+		/// If any of the following are null:
+		/// <see cref="BindableProperty.PropertyName"/>
+		/// <see cref="BindableProperty.ReturnType"/>
+		/// <see cref="BindableProperty.DeclaringType"/>
+		/// </exception>
+		/// <exception cref="ArgumentException">
+		/// If the type of the <see cref="BindableProperty.DefaultValue"/>
+		/// and <see cref="BindableProperty.ReturnType"/> don't match.
+		/// </exception>
+		public BindableProperty Build()
+		{
+			CheckParameter(!string.IsNullOrEmpty(bpParameters.PropertyName), nameof(bpParameters.PropertyName));
+			CheckParameter(bpParameters.ReturnType != default, nameof(bpParameters.ReturnType));
+			CheckParameter(bpParameters.DeclaringType != default, nameof(bpParameters.DeclaringType));
```

Comment:
```csharp
if (string.IsNullOrEmpty(bpParameters.PropertyName)
    throw new ArgumentNullException(parameterName, $"{parameterName} is required");

if (bpParameters.ReturnType == null)
    throw new ArgumentNullException(parameterName, $"{parameterName} is required");
```

Do you think this approach is better?

Reply:
For an OSS project like this I think it is. The code clearly explains what it is doing where the method name `CheckParameter()` doesn't really explain what is happening. 

When I first glanced at this file I didn't realize the `CheckParameter()` method was throwing an exception and it was hiding logic. It may just be how I read the code file, but if I didn't notice it right away, I'm sure others would have the same problem. I think what you suggested will make this easier to maintain

Reply:
I agree with @ahoefling - I greatly prefer the more verbose and explicit syntax of test-and-throw instead of tucking it away in a method.